### PR TITLE
Add an Export locality to hints

### LINKED ISSUE #11792: "Global" locality for hints
When defining hints in stdlib2, we'd like to give them a "global" locality behavior (as opposed to "superglobal"). I believe there's no way to do that today. The related `Loose Hint Behavior` controls the behavior when using hints, not when defining then, as discussed in https://github.com/coq/stdlib2/issues/6.

Cc @ppedrot

### PR DESCRIPTION
**Pull request by ppedrot**
We add an intermediate locality for hints called `export`, whose name is inspired by the feature for options which shares the same behaviour. Such hints are only made available when the module that contains them is imported, not merely required.

The locality is currently only available as an attribute, even though we may introduce an option to change the default behaviour from `global` to `export` .

Fixes / closes #11792.

- [X] Added / updated test-suite
- [X] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

Overlays:
- https://github.com/mattam82/Coq-Equations/pull/267
- https://github.com/mit-plv/fiat/pull/35

### CODE REVIEW

**Comment by ppedrot** [comment]
@Zimmi48 you mentioned in #11665 that `Export` should be turned into a legacy attribute, maybe the correct infrastructure coud be done in this PR since it's relying on its availability as a proper attribute.

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]: Or maybe it is this additional space and the one below.

Anyway, I suggest that you remove entirely this command variant and its description.

**Review by maximedenes — COMMENTED** [review]
Very nice! I already use this with success in stdlib2.

Implementationwise, do you think it would be doable to push the three-valued locality a bit farther? I.e. through `add_resolves` and the likes.

**Comment by ppedrot** [comment]
@maximedenes I have a PR that removes most of this API, so I don't think it's worth removing it for the time being.

**Review by Zimmi48 — COMMENTED** [review]
> @Zimmi48 you mentioned in #11665 that `Export` should be turned into a legacy attribute, maybe the correct infrastructure coud be done in this PR since it's relying on its availability as a proper attribute.

There's a risk of interaction between the two PRs if this one goes further than this without the other being merged.

File: tactics/hints.ml
Comment (gares) [inline tactics/hints.ml]: I think we are on 4.03+ now, we should start naming fields instead of using comments
`AddHints of { superglobal : bool; hints : hint_entry list }` should do the trick

Comment (ppedrot) [reply tactics/hints.ml]: Done.

File: doc/sphinx/proof-engine/tactics.rst
Comment (ppedrot) [reply doc/sphinx/proof-engine/tactics.rst]: The doc failure seems to be due to the missing definition of the attribute `export`, instead.

File: doc/sphinx/proof-engine/tactics.rst
Comment (ppedrot) [reply doc/sphinx/proof-engine/tactics.rst]: Backwards compat maybe... We should at least deprecate them at some point.

**Comment by ppedrot** [comment]
@Zimmi48 should I creat an attr entry in the Attributes section of gallina-specification-language.rst, alongside with the local / global / whatnot?

**Comment by Zimmi48** [comment]
> @Zimmi48 should I creat an attr entry in the Attributes section of gallina-specification-language.rst, alongside with the local / global / whatnot?

Please put it at the bottom of this section: https://coq.github.io/doc/master/refman/proof-engine/vernacular-commands.html#coq:cmd.global

**Comment by Zimmi48** [comment]
I'll wait for this PR to be merged to finish #11665 and make `Export` a legacy attribute as well.

**Comment by ppedrot** [comment]
@Zimmi48 this will not define the attribute though, and thus won't fix the compilation errors in the refman...

**Comment by ppedrot** [comment]
Or you mean bluntly adding this as an attribute entry, even if this makes the documentation quite of clunky for the moment?

**Comment by Zimmi48** [comment]
> Or you mean bluntly adding this as an attribute entry, even if this makes the documentation quite of clunky for the moment?

Yes. Add `.. attr:: export` plus a short description, and I will refine this in #11665.

**Comment by ppedrot** [comment]
CI failure is spurious, bedrock is broken in master due to some OOM.

**Review by Zimmi48 — APPROVED** [review]
LGTM
Regarding the behavior of attributes inside sections, at least a warning could be emitted when using `global` or `export` that they have no effect.

**Comment by ppedrot** [comment]
This is ready and needs an assignee.

**Comment by Zimmi48** [comment]
WDYT of my suggestion?

**Comment by ppedrot** [comment]
@Zimmi48 I don't really want to touch the behaviour of the command with other attributes here, but I can open another PR if needed.

File: plugins/ltac/g_auto.mlg
Comment (Zimmi48) [inline plugins/ltac/g_auto.mlg]: Is it normal that `Hint Cut` does not accept the new `export` attribute?

Comment (ppedrot) [reply plugins/ltac/g_auto.mlg]: The whole hint registering system is bonkers. It needs a complete rewriting with a preliminary serious design phase for the semantics of the commands. This PR was born of the urgency of covering 95% of the modularity-breaking code, and even though I think this one can be handled as hackishly as we are doing it for standard hints, I am not too confident it will behave that well...

Comment (Zimmi48) [reply plugins/ltac/g_auto.mlg]: Since the affected commands are `Hint Resolve`, `Hint Immediate`, `Hint Unfold` and `Hint Extern`, it would be better to clearly document that these are the four commands that accept the `export` attribute and avoid a new "documentation is lying" bug report. I guess extending the support to `Hint Mode`, `Hint Cut`, `Hint Transparent` and `Hint Opaque` is up to @mattam82.

File: vernac/obligations.ml
Comment (Zimmi48) [inline vernac/obligations.ml]: What about renaming this one `add_local_hint`?

File: vernac/vernacentries.ml
Comment (Zimmi48) [inline vernac/vernacentries.ml]:
```suggestion
      "This command does not support the global attribute in sections.");
```

Also, not sure this should be an error given that the usual unsupported attribute message is just a fatal-by-default warning (OK this was already an error message so you don't want to change this).

Also, why is this check only performed for `global` and not for `export`?

Finally, this was the only place where the `enforce_module_locality` function was used, so please remove it from `vernac/locality.ml{,i}`.

File: plugins/ltac/g_auto.mlg
Comment (ppedrot) [reply plugins/ltac/g_auto.mlg]: Done.

**Review by Zimmi48 — COMMENTED** [review]
I reviewed the rest of this PR. Looks OK to me as a first step.

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]:
```suggestion
     the :cmd:`Hint Resolve`, :cmd:`Hint Immediate`, :cmd:`Hint Unfold` and :cmd:`Hint Extern` variants of
```

**Review by Zimmi48 — COMMENTED** [review]
What happens if I write `#[ export ] Hint Transparent foo.`? Is it silently ignored and interpreted as global?

**Comment by Zimmi48** [comment]
It seems that it wouldn't be much more effort to make `Instance` support `export` as well but OK to leave this as a further step too.

File: vernac/obligations.ml
Comment (ppedrot) [reply vernac/obligations.ml]: Done.

File: vernac/vernacentries.ml
Comment (ppedrot) [reply vernac/vernacentries.ml]: Tweaked the message, and added the check for export as well. I didn't remove `enforce_module_locality` though, as it might be used by plugins, and furthermore it would make the Locality API kind of asymmetrical, since all other localities come as pairs `make` / `enforce`. I'd rather remove this API at once instead.

File: doc/sphinx/proof-engine/tactics.rst
Comment (ppedrot) [reply doc/sphinx/proof-engine/tactics.rst]: Done.

**Comment by ppedrot** [comment]
> Is it silently ignored and interpreted as global?

Now it fails properly.

File: vernac/vernacentries.ml
Comment (Zimmi48) [reply vernac/vernacentries.ml]: OK that makes sense.

**Review by Zimmi48 — APPROVED** [review]
LGTM. I will merge this later today or tomorrow morning.

File: plugins/ltac/g_auto.mlg
Comment (Blaisorblade) [reply plugins/ltac/g_auto.mlg]: 🎉! Wouldn't instances fall in the 95%? But I'm happy with an issue :-)

**Pull request merged**
